### PR TITLE
Archive info database overhaul

### PR DESCRIPTION
**Pull request by EvilDrPurple**
#### Fixes

* Part of https://github.com/Police-Data-Accessibility-Project/data-sources-app/issues/309 and https://github.com/Police-Data-Accessibility-Project/automatic-archives/issues/21

#### Description

* This PR sets up the infrastructure for a new table called `data_sources_archive_info` that contains `update_frequency`, `last_cached` and a new column called `next_cache`
* The following database query will be used to setup the new table and transfer the data over: [query.txt](https://github.com/user-attachments/files/16339202/query.txt)

#### Testing

1. Run the query in a freshly wiped sandbox database
2. Checkout the branch locally and setup the v2 testing environment
3. Run `pytest` in the tests folder
4. Profit

### CODE REVIEW

File: database_client/constants.py
Comment (maxachis) [inline database_client/constants.py:1]: Smart to make a variable. We do need to figure out how to reorganize these constants, but that's beyond the scope of this PR.

File: middleware/archives_queries.py
Comment (maxachis) [inline middleware/archives_queries.py:57]: I'm interested in the removal of this else conditional -- if I recall correctly, a broken url should not be archived, correct? I may be wrong on that (and am happy to be corrected if that's the case). That being said, such a source could still count as being "archived", but it's a tad ambiguous. Could you clarify your reasoning with this decision?

File: resources/Archives.py
Comment (maxachis) [inline resources/Archives.py:1]: I recall you giving me your rationale for rendering these as DateTimes rather than dates, so nice work here.

File: tests/integration/test_archives.py
Comment (maxachis) [inline tests/integration/test_archives.py:71]: Converting to a more vertical orientation here is much appreciated.

File: tests/test_database_client.py
Comment (maxachis) [inline tests/test_database_client.py:215]: Yeah, it does make more sense to have the `strftime` method alongside everything else, rather than down at the bottom. Good catch 🥅

File: utilities/common.py
Comment (maxachis) [inline utilities/common.py:10]: If we're going to have certain keys follow specific date formats, I'd rather the function be generalized rather than having a condition for a single key that is only relevant in one of the (many) places this function is called. I could see an optional secondary dictionary where the key is the data_dict key and the value is the desired strftime format for that. 

I won't block if we don't resolve this now, but if we don't, I would want a minor issue made so that we can address it later on.

**Review by maxachis — CHANGES_REQUESTED** [review]
Overall, it looks good, and the tests run successfully. I do have some questions on design decisions that I'd like clarified/resolved before merging successfully. I'll also hold off on merging https://github.com/Police-Data-Accessibility-Project/prod-to-dev-migration/pull/19#pullrequestreview-2192777670 (despite it being approved) just because I think these should be a package deal.

File: database_client/constants.py
Comment (EvilDrPurple) [reply database_client/constants.py:1]: Yeah for sure

File: database_client/dynamic_query_constructor.py
Comment (EvilDrPurple) [reply database_client/dynamic_query_constructor.py:144]: I did use it on line 74, concating it with `data_sources_columns`

File: middleware/archives_queries.py
Comment (EvilDrPurple) [reply middleware/archives_queries.py:57]: Okay this is going to be a little complicated to explain so let me know if I'm explaining it incoherently:
If you look at what was previously line 451 of `database_client.py` in `update_url_status_to_broken()` I had to remove `"last_cached": last_cached,` as part of the data package being sent to `update_data_source()` since `last_cached` is no longer part of the `data_sources` table. Since `last_cached` was updated in both the `if` and `else` branches, removing the `else` means that `last_cached` will be updated regardless of if it is broken or not, just like previously.

File: tests/integration/test_archives.py
Comment (EvilDrPurple) [reply tests/integration/test_archives.py:71]: Yeah it was running off the screen from all the things I added haha

File: tests/test_database_client.py
Comment (EvilDrPurple) [reply tests/test_database_client.py:215]: I believe I did it this way since `datetime` in Postgres will accept milliseconds and I didn't want to deal with all that in there!

File: utilities/common.py
Comment (EvilDrPurple) [reply utilities/common.py:10]: Gotcha! I didn't like doing it this way either haha. I figured maybe we could convert all `date` columns in the database to `datetime` as another solution, so that one `if` statement would work. But that seemed outside of scope

File: middleware/archives_queries.py
Comment (EvilDrPurple) [reply middleware/archives_queries.py:57]: But I was also a little confused why it was set up this way in the first place. Why, when a link is broken do we update the `last_cached` column? Surely the source has not been archived if it is broken. However, I didn't want to change the logic unless you believe it should be now or in the future

File: database_client/dynamic_query_constructor.py
Comment (maxachis) [reply database_client/dynamic_query_constructor.py:144]: Well jeez! Flake8 lied to me! No issue here, then.

File: middleware/archives_queries.py
Comment (maxachis) [reply middleware/archives_queries.py:57]: Well, I don't know the original design decision in this case, but in my thinking there is a possibility that a link was _previously_ operational and then archived, so last_cached might still be applicable in that context. But it is also a case where I think we could clarify the design for this later on.

Comment (maxachis) [reply middleware/archives_queries.py:57]: At any rate, you've convinced me with this! _**RESOLVED**_.

File: utilities/common.py
Comment (maxachis) [reply utilities/common.py:10]: I think that's an excellent idea, and probably a better choice -- I can't see any cases where we want date over datetime. If you make an issue for that (and for resolving this as well) I'll consider this resolved and will be happy to merge.

**Pull request merged**
